tests/integration/test_dev_server.py

```diff
@@ -0,0 +1,156 @@
+import asyncio
+import nikola.plugins.command.auto as auto
+from nikola.utils import get_logger
+import pytest
+import pathlib
+import socket
+from typing import Optional
+import urllib.request
+
+from ..helper import FakeSite
+
+SERVER_ADDRESS = "localhost"
+TEST_MAX_DURATION = 10  # Watchdog: Give up the test if it did not succeed during this time span.
+SERVER_STARTUP_TIME = 0.1  # Max time, in seconds, we expect the server to need to come up.
+
+# Folder that has the fixture file we expect the server to serve:
+OUTPUT_FOLDER = pathlib.Path.cwd() / "tests" / "data" / "dev_server_sample_output_folder"
+
+LOGGER = get_logger("test_dev_server")
+
+
+def find_unused_port() -> int:
+    """Ask the OS for a currently unused port number.
+
+    (More precisely, a port that can be used for a TCP server servicing SERVER_ADDRESS.)
+    We use a method here rather than a fixture to minimize side effects of failing tests.
+    """
+    s = socket.socket()
+    try:
+        ANY_PORT = 0
+        s.bind((SERVER_ADDRESS, ANY_PORT))
+        address, port = s.getsockname()
+        LOGGER.info("Trying to set up dev server on http://%s:%i/", address, port)
+        return port
+    finally:
+        s.close()
+
+
+def test_serves_root_dir(site, expected_text) -> None:
+    command_auto = auto.CommandAuto()
+    command_auto.set_site(site)
+    options = {
+        "browser": False,
+        "ipv6": False,
+        "address": SERVER_ADDRESS,
+        "port": find_unused_port(),
+        "db-file": "/dev/null",
+        "backend": "No backend",
+        "no-server": False
+    }
+
+    # We start an event loop, run the test in an executor,
+    # and wait for the event loop to terminate.
+    # These variables help to transport the test result to
+    # the main thread outside the event loop:
+    test_was_successful = False
+    test_problem_description = "Async test setup apparently broken"
+    test_inner_error: Optional[BaseException] = None
+
+    async def grab_loop_and_run_test() -> None:
+        loop = asyncio.get_running_loop()
+        watchdog_handle = loop.call_later(TEST_MAX_DURATION, lambda: loop.stop())
+        nonlocal test_problem_description
+        test_problem_description = f"Test did not complete within {TEST_MAX_DURATION} seconds."
+
+        def run_test() -> None:
+            nonlocal test_was_successful, test_problem_description, test_inner_error
+            try:
+                server_base_uri = f"http://{options['address']}:{options['port']}/"
+
+                # First subtest: Does the dev server serve the fixture HTML file as expected?
+                with urllib.request.urlopen(server_base_uri) as res:
```

Comment:
You can (should?) use `requests` instead of `urllib.request` for cleaner and clearer code. (`requests` is a mandatory dependency of Nikola.)

Reply:
That's good news! I should have checked whether you use it.